src/librepgp/stream-common.cpp

```diff
@@ -210,17 +211,24 @@ src_skip(pgp_source_t *src, size_t len)
 
     if (len < sizeof(sbuf)) {
         return src_read(src, sbuf, len);
-    } else {
-        buf = calloc(1, len);
-
-        if (buf == NULL) {
-            return -1;
-        } else {
-            res = src_read(src, buf, len);
+    }
+    
+    buf = calloc(1, std::min((size_t) PGP_INPUT_CACHE_SIZE, len));
+    if (!buf) {
+        return -1;
+    }
+    
+    while (len > 0) {
+        ssize_t opres = src_read(src, buf, std::min((size_t) PGP_INPUT_CACHE_SIZE, len));
+        if (opres < 0) {
             free(buf);
-            return res;
+            return opres;
         }
+        res += opres;
+        len -= opres;
```

Comment:
Omg, so easy. Thanks! Will fix this once get back to computer.

Reply:
I'm thinking the issue may lie elsewhere. `src_read` keeps returning `0` in this loop, so it ends up infinite.

Reply:
@dewyatt actually we should also check for src_eof(), on end of stream code will not return -1 but 0. And this may happen, for instance, on indeterminate stream length or cleartext signed stream.